src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1011 @@
+            }
+
+            int windowUpdate = (int)((uint)((_incomingBuffer.ReadBytes[0] << 24) | (_incomingBuffer.ReadBytes[1] << 16) | (_incomingBuffer.ReadBytes[2] << 8) | _incomingBuffer.ReadBytes[3]) & 0x7FFFFFFF);
+
+            Debug.Assert(windowUpdate >= 0);
+            if (windowUpdate == 0)
+            {
+                throw new Http2ProtocolException(Http2ProtocolErrorCode.ProtocolError);
+            }
+
+            // TODO: Window accounting
+
+            _incomingBuffer.Consume(frameHeader.Length);
+        }
+
+        private void WriteFrameHeader(FrameHeader frameHeader)
+        {
+            _outgoingBuffer.EnsureWriteSpace(FrameHeader.Size);
+            frameHeader.Write(_outgoingBuffer.WriteBytes);
+            _outgoingBuffer.Commit(FrameHeader.Size);
+        }
+
+        private enum FrameType : byte
+        {
+            Data = 0,
+            Headers = 1,
+            Priority = 2,
+            RstStream = 3,
+            Settings = 4,
+            PushPromise = 5,
+            Ping = 6,
+            GoAway = 7,
+            WindowUpdate = 8,
+            Continuation = 9
+        }
+
+        private struct FrameHeader
+        {
+            public readonly int Length;
+            public readonly FrameType Type;
+            public readonly FrameFlags Flags;
+            public readonly int StreamId;
+
+            public const int Size = 9;
+            public const int MaxLength = 16384;
+
+            public const int PriorityInfoLength = 5;       // for both PRIORITY frame and priority info within HEADERS
+            public const int PingLength = 8;
+            public const int WindowUpdateLength = 4;
+
+            public FrameHeader(int length, FrameType type, FrameFlags flags, int streamId)
+            {
+                Debug.Assert(length <= MaxLength);
+                Debug.Assert(streamId >= 0);
+
+                Length = length;
+                Type = type;
+                Flags = flags;
+                StreamId = streamId;
+            }
+
+            public bool PaddedFlag => (Flags & FrameFlags.Padded) != 0;
+            public bool AckFlag => (Flags & FrameFlags.Ack) != 0;
+            public bool EndHeadersFlag => (Flags & FrameFlags.EndHeaders) != 0;
+            public bool EndStreamFlag => (Flags & FrameFlags.EndStream) != 0;
+            public bool PriorityFlag => (Flags & FrameFlags.Priority) != 0;
+
+            public static FrameHeader Read(ReadOnlySpan<byte> buffer)
+            {
+                Debug.Assert(buffer.Length >= Size);
+
+                return new FrameHeader(
+                    (buffer[0] << 16) | (buffer[1] << 8) | buffer[2],
+                    (FrameType)buffer[3],
+                    (FrameFlags)buffer[4],
+                    (int)((uint)((buffer[5] << 24) | (buffer[6] << 16) | (buffer[7] << 8) | buffer[8]) & 0x7FFFFFFF));
+            }
+
+            public void Write(Span<byte> buffer)
```

Comment:
Nit: I'm confused by these method names.  Generally a method that's called Write and that takes a span/array/etc. as an argument writes that span/array/etc. to the instance, but this is actually doing the reverse, writing the frame to the argument?  If so, I'd suggest changing the name a bit, e.g. Write => WriteTo, or CopyTo, or something like that.

Reply:
I've changed to "ReadFrom" and "WriteTo"